### PR TITLE
Davutkarabay/reversal workflow 2

### PR DESCRIPTION
**Pull request by nitishm**
Fork from #194

### CODE REVIEW

File: api/v1alpha1/appgroup_types.go [hunk 1]
```diff
@@ -262,10 +262,16 @@ func (in *ApplicationGroup) Progressing() {
 // RollingBack sets the meta.ReadyCondition to 'True' and
 // meta.RollingBack reason and message
 func (in *ApplicationGroup) RollingBack() {
```
Comment (jonathan-innis) [inline api/v1alpha1/appgroup_types.go:264]: Just to clarify, what is the difference now between these two conditions?

File: pkg/meta/conditions.go [hunk 2]
```diff
@@ -30,6 +30,11 @@ const (
 	// ReleasedCondition represents the status of the last release attempt
 	// (install/upgrade/test) against the latest desired state.
 	ReleasedCondition string = "Released"
+
+	// ReversingCondition is the name of the Reversed condiiton
```
Comment (davutkarabay) [inline pkg/meta/conditions.go:34]: nit: typo "condiition"

**Comment by jonathan-innis** [comment]
> apiVersion: apiextensions.k8s.io/v1

Not sure why this got updated, I see that lines got wrapped to not be as long, can we duplicate this change in the `chart/orkestra/crd` as well?

---
Refers to: config/crd/bases/orkestra.azure.microsoft.com_applicationgroups.yaml:3 in cbd33f3. [](commit_id = cbd33f3debdb6b939921181dc2528072a2580b08, deletion_comment = False)

File: controllers/appgroup_controller.go [hunk 3]
```diff
@@ -31,6 +33,7 @@ import (
 const (
```
Comment (jonathan-innis) [inline controllers/appgroup_controller.go:33]: Typically the standard is to move finalizers/annotations related to the CR in the `appgroup_types` file

File: api/v1alpha1/appgroup_types.go (hunk 1 above)
Comment (nitishm) [reply api/v1alpha1/appgroup_types.go:264]: RollingBack is on upgrade failures where we rollback to previous successful spec and Reversing is for installation failures
Not sure if that was the question 🤔

File: controllers/appgroup_controller.go [hunk 4]
```diff
@@ -499,32 +508,89 @@ func (r *ApplicationGroupReconciler) cleanupWorkflow(ctx context.Context, logr l
...
 				}
+
+				err = r.Client.Delete(ctx, &wf)
+				if err != nil {
+					logr.Error(err, "failed to delete workflow CRO - continuing with cleanup")
+					return
+				}
+
+				// reverse workflow started - requeue
+				requeue = true
```
Comment (jonathan-innis) [inline controllers/appgroup_controller.go:554]: Can we be explicit with our return statements, setting the variable and then returning like this almost hides the behavior a bit in my opinion

File: pkg/meta/conditions.go [hunk 5]
```diff
@@ -30,6 +30,11 @@ const (
...
+	// This captures the state of reversing the forward workflow
+	// on deletion of the ApplicationGroup
+	ReversingCondition string = "Reversing"
```
Comment (jonathan-innis) [inline pkg/meta/conditions.go:37]: Why do we have both a reversing condition and a reversing reason? It seems that we are only using the `ReversingReason` and not the `ReversingCondition`?

File: pkg/workflow/argo.go [hunk 6]
```diff
@@ -239,17 +353,17 @@ func (a *argo) generateAppGroupTpls(ctx context.Context, g *v1alpha1.Application
 	if err != nil {
 		return fmt.Errorf("failed to generate application DAG templates : %w", err)
 	}
-	a.updateWorkflowTemplates(adt...)
+	updateWorkflowTemplates(a.wf, adt...)
```
Comment (jonathan-innis) [inline pkg/workflow/argo.go:356]: Same comment as before, why this change?

File: api/v1alpha1/appgroup_types.go (hunk 1 above)
Comment (jonathan-innis) [reply api/v1alpha1/appgroup_types.go:264]: I see, is there a reason we have a distinction between these two? Aren't they essentially the same action?

Comment (jonathan-innis) [reply api/v1alpha1/appgroup_types.go:264]: My personal opinion is that we just use reversing reason as universal for both rollback and for delete

File: pkg/meta/conditions.go (hunk 5 above)
Comment (nitishm) [reply pkg/meta/conditions.go:37]: I wasnt sure how this worked lol. So should I remove ReversingCondition?

File: api/v1alpha1/appgroup_types.go (hunk 1 above)
Comment (nitishm) [reply api/v1alpha1/appgroup_types.go:264]: Hmm - we do some special handling for RollingBack early in the reconciliation - like restoring the lastSuccessfulSpec and re-applying the Workflow generated from it. 
So we need to handle the events separately. Does that make sense?

File: pkg/workflow/argo.go [hunk 7]
```diff
@@ -539,6 +653,28 @@ func defaultExecutor(timeout time.Duration) v1alpha12.Template {
 	}
 }
 
+func defaultReverseExecutor() v1alpha12.Template {
```
[the remark this answers is not in the document]
Comment (nitishm) [reply pkg/workflow/argo.go:656]: That is the plan - but it will happen in a future PR

File: pkg/workflow/argo.go (hunk 6 above)
Comment (nitishm) [reply pkg/workflow/argo.go:356]: @davutkarabay can answer that.

File: controllers/appgroup_controller.go (hunk 3 above)
Comment (nitishm) [reply controllers/appgroup_controller.go:33]: Will move them in a future commit / PR. This would require the ApplicationGroup finalizer to be moved as well.

File: pkg/workflow/argo.go (hunk 6 above)
Comment (davutkarabay) [reply pkg/workflow/argo.go:356]: Previously the function updated the forward workflow with the templates. Now we are using it to update either the forward workflow or the reverse workflow which is passed as parameter.

File: api/v1alpha1/appgroup_types.go (hunk 1 above)
Comment (jonathan-innis) [reply api/v1alpha1/appgroup_types.go:264]: I see, I still think reversing should be moved from the ReadyCondition to DeployCondition

File: pkg/meta/conditions.go (hunk 5 above)
Comment (jonathan-innis) [reply pkg/meta/conditions.go]: Yes, we are not using this, so just the reason is fine

File: api/v1alpha1/appgroup_types.go (hunk 1 above)
Comment (jonathan-innis) [reply api/v1alpha1/appgroup_types.go:264]: This may need to be a longer conversation about what each condition should be capturing

File: pkg/workflow/argo.go (hunk 6 above)
Comment (jonathan-innis) [reply pkg/workflow/argo.go:356]: Isn't the end result still the same or am I missing something?

**Review by jonathan-innis — CHANGES_REQUESTED** [review]

File: pkg/workflow/argo.go (hunk 6 above)
Comment (nitishm) [reply pkg/workflow/argo.go:356]: In the previous iteration, we only had the forward workflow (`a.wf`) member in the Argo type struct. So it made sense to make this a method and not a standalone function. But since we have added the reverse workflow as a member of Argo struct, it made sense to make this a function instead of an associate method. This is because we leverage this function both when building the forward workflow and also the reverse workflow

**Review by davutkarabay — DISMISSED** [review]

File: pkg/workflow/argo.go (hunk 6 above)
Comment (davutkarabay) [reply pkg/workflow/argo.go:356]: @jonathan-innis do you see the usage of updateWorkflowTemplates() in lines #326 and #328 in this file (end of generateReverseWorkflow()). Does this clarify?

Comment (jonathan-innis) [reply pkg/workflow/argo.go:356]: ah, I see one letter and I missed it, all good then 🚀

**Comment by nitishm** [comment]
@jonathan-innis We good to merge?

File: controllers/appgroup_controller.go [hunk 8]
```diff
@@ -31,6 +33,7 @@ import (
 const (
 	appgroupNameKey                   = "appgroup"
 	finalizer                         = "application-group-finalizer"
+	rwffinalizer                      = "reverse-workflow-finalizer"
```
Comment (jonathan-innis) [inline controllers/appgroup_controller.go:36]: Can we discuss again why we are using a finalizer? Why not just execute delete when we have confirmed that the workflow has completed? The main reason I am raising this is because it block all delete of the workflow CR. This is problematic because `helm uninstall` will not be able to remove the CRO.

**Review by jonathan-innis — CHANGES_REQUESTED** [review]

File: controllers/appgroup_controller.go (hunk 8 above)
Comment (jonathan-innis) [reply controllers/appgroup_controller.go:36]: Other than this, nice work @davutkarabay 🚀  looks great!

Comment (nitishm) [reply controllers/appgroup_controller.go:36]: > Can we discuss again why we are using a finalizer? Why not just execute delete when we have confirmed that the workflow has completed? The main reason I am raising this is because it block all delete of the workflow CR. This is problematic because `helm uninstall` will not be able to remove the CR.

This is mostly for better UX. We would like to show the end-user both the forward workflow (that is being reversed) and the reverse workflow on the dashboard. Can we have the hook look for finalizers and remove them during uninstall? Can we think of a reason why it may or may not be a good idea?

Comment (jonathan-innis) [reply controllers/appgroup_controller.go:36]: Why not delete both after the reverse succeeds? How is this any different than calling the delete before with a finalizer?

Comment (nitishm) [reply controllers/appgroup_controller.go:36]: Sure - no reason why we shouldnt do that. Probably easier to follow that way. Let me take a stab at it.

Comment (jonathan-innis) [reply controllers/appgroup_controller.go:36]: Thanks, this would simplify things and reduce conflicts :)

File: controllers/appgroup_controller.go [hunk 9]
```diff
@@ -499,32 +506,66 @@ func (r *ApplicationGroupReconciler) cleanupWorkflow(ctx context.Context, logr l
...
+			logr.Error(err, "failed to GET workflow object with an unrecoverable error")
+		} else {
+			// check the completion of the reverse workflow
+			if !rwf.Status.FinishedAt.IsZero() {
+				logr.Info("reverse workflow is finished")
 
-		for _, bucket := range rev {
-			for _, hr := range bucket {
-				err = r.Client.Delete(ctx, &hr)
+				err = r.Client.Delete(ctx, &wf)
```
Comment (jonathan-innis) [inline controllers/appgroup_controller.go:543]: Don't we also need to delete the reverse workflow?

**Review by jonathan-innis — CHANGES_REQUESTED** [review]

File: controllers/appgroup_controller.go (hunk 9 above)
Comment (nitishm) [reply controllers/appgroup_controller.go:543]: Not needed since it is owned by the forward workflow using the ownersReference field. So it will get garbage collected on deleting the forward workflow object

**Review by jonathan-innis — APPROVED** [review]
🚀
